subprojects/core/src/main/java/org/gradle/api/internal/plugins/DslObject.java

```diff
@@ -79,6 +86,26 @@ public Class getDeclaredType() {
         return object.getClass();
     }
 
+    public TypeOf<Object> getPublicType() {
```

Comment:
this should be merged into `getDeclaredType()` (or vice versa). They're the same thing.

Reply:
If I get rid of `getDeclaredType`, would I change users to work with `TypeOf`? 

Or if I change `getPublicType` to return a `Class` instead (and get rid of `getDeclaredType`), what do we do for things that implement `HasPublicType` and return a `TypeOf`?  Would I just extract the "raw" class from that?

Reply:
We should keep `TypeOf<?>` declarations. It's the only way to capture generics properly.